wallet-new/src/Cardano/Wallet/API/V1/Errors.hs

```diff
@@ -1,19 +1,104 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
 
 module Cardano.Wallet.API.V1.Errors where
 
 import           Universum
 
-import           Cardano.Wallet.API.V1.Types (WalletError (..))
-import           Data.Aeson (encode)
+import           Data.Aeson
+import           Data.Aeson.TH (deriveJSON)
 import qualified Network.HTTP.Types.Header as HTTP
 import           Servant
+import           Test.QuickCheck (Arbitrary (..), oneof)
 
+import           Cardano.Wallet.API.V1.TH (conNamesList, deriveWalletErrorJSON)
+
+--
+-- Error handling
+--
+
+-- | Type representing any error which might be thrown by wallet.
+--
+-- Errors are represented in JSON in the following way:
+-- ```
+-- {
+--     "message" : <constr_name>,
+--     "diagnostic" : <data>
+-- }
+-- ```
+-- where `<constr_name>` is a string containing name of error's constructor (e. g. `NotEnoughMoney`),
+-- and `<data>` is an object containing additional error data.
+-- Additional data contains constructor fields, field names are record field names without
+-- a `we` prefix, e. g. for `OutputIsRedeem` error "diagnostic" field will be the following:
+-- ```
+-- {
+--     "address" : <address>
+-- }
+-- ```
+-- If constructor does not define record field names, then data values are labeled with
+-- keys "v1", "v2" and so on, e. g. for `NotRecordError` error:
+-- ```
+-- {
+--     "v1" : 3,
+--     "v2" : "blah"
+-- }
+-- ```
+-- If constructor does not have additional data (like in case of `WalletNotFound` error),
+-- then "diagnostic" field will be empty object.
+--
+-- TODO: change fields' types to actual Cardano core types, like `Coin` and `Address`
+data WalletError =
+      NotEnoughMoney { weNeedMore :: !Int }
+    | OutputIsRedeem { weAddress :: !Text }
+    | SomeOtherError { weFoo :: !Text, weBar :: !Int }
+    | NotRecordError !Int !Text
+    | MigrationFailed { weDescription :: !Text }
+    | WalletNotFound
+    deriving (Show, Eq, Generic)
+
+--
+-- Instances for `WalletError`
+
+deriveWalletErrorJSON ''WalletError
+
+instance Exception WalletError where
+
+-- TODO: generate `Arbitrary` instance with TH too?
+instance Arbitrary WalletError where
+    arbitrary = oneof
+        [ NotEnoughMoney <$> arbitrary
+        , OutputIsRedeem <$> pure "address"
+        -- , SomeOtherError <$> pure "blah" <*> arbitrary
+        , NotRecordError <$> arbitrary <*> pure "blah"
+        , MigrationFailed <$> pure "migration"
+        , pure WalletNotFound
+        ]
+
+--
+-- Helpers
+--
+
+-- | List of all existing error tags. Populates automatically
+allErrorsList :: [Text]
+allErrorsList = $(conNamesList ''WalletError)
```

Comment:
Are we showing this list in our Swagger spec? I think the key point in design we agreed upon is that we want not only to show the list of errors, but a nicely-formatted table which maps the error to the code and its description. My dream table to attach to to top of the swagger documentation would look like this:

| Error Name      | HTTP Error code | Description                                 |
|-----------------|-----------------|---------------------------------------------|
| WalletNotFound  | 404             | "The requested wallet cannot be found."     |
| MigrationFailed | 422             | "Migration between V0 and V1 types failed." |

Do you think that's possible with the current design? I'm happy if the description includes fake data, for example when rendering a `NotEnoughMoney` error, I'm fine with generating an arbitrary instance which would populate the description accordingly.

Reply:
I can add function similar to `walletHTTPError` (like `walletErrorDescription`) to provide error description for every type of error, so we are able to construct such table.

Note on these functions: I don't particularly like this approach (it looks like boilerplate, and compiler don't prevent a developer from forgetting to provide an HTTP error or description for newly added error constructor if `-WError` isn't set), but I don't see more elegant way to do this.